### PR TITLE
[Solid] Transfer custom solvers to SolversApp

### PR DESCRIPTION
**Pull request by josep-m-carbonell**
SolidApp strategies, schemes,  builders and linear solvers, now located in SolversApp, also python solvers scripts. 
The application is added to appveyor because now Solid app depends on it.

### CODE REVIEW

**Comment by philbucher** [comment]
Hm I am not sure if it is a good idea to overload appveyor even further, it is already out-of-time in many PRs

**Comment by loumalouomega** [comment]
Why not adding this to the ExternalSolverApplication?, instead of creating a duplicated application for solvers?

**Comment by RiccardoRossi** [comment]
I agree with @philbucher 

furthermore we did an effort to centralize this sort of things in the ExternalSolversApplication so i really think we should not replicate it elsewhere

**Review by RiccardoRossi — CHANGES_REQUESTED** [review]
also blocking since this changes 1550 files

**Comment by pooyan-dadvand** [comment]
@josep-m-carbonell I would strongly suggest you to divide this PR to several smaller ones. It is impossible to be reviewed as it is and it won't be incorporated in the release without review. 

I also support the idea of centralizing solvers in external solvers application.

**Comment by josep-m-carbonell** [comment]
The only changes that can affect to other applications not depending on SolidMechanicsApplication is in the common CMakeList.txt of the applications, to include SolverApplication. And the addition to appveyor.

I don't know if the addition to appveyor will need more compilation time or not, because files form SolidMechanicsApp are just transferred to SolversApp. It must be more or less the same compilation time.

**Comment by josep-m-carbonell** [comment]
`... blocking since this changes 1550 files`

This is a new application including some bsd libraries, so, yes there are a lot of new files.

`It is impossible to be reviewed as it is and it won't be incorporated in the release without review.`

If you don't want or you don't have time to review it, some other people will do it. The PR can not be splitted.

**Comment by RiccardoRossi** [comment]
i am adding for discussion to @KratosMultiphysics/technical-committee

**Comment by RiccardoRossi** [comment]
however, could u explain why you cannot depend on the ExternalSolversApplication ?

**Comment by loumalouomega** [comment]
> `... blocking since this changes 1550 files`
> 
> This is a new application including some bsd libraries, so, yes there are a lot of new files.
> 
> `It is impossible to be reviewed as it is and it won't be incorporated in the release without review.`
> 
> If you don't want or you don't have time to review it, some other people will do it. The PR can not be splitted.

All PR can be splitted, I have done this many times, even if some PR depend of other PR. For example you can do a PR only for the solvers. Solvers aren't supposed to depend of anything else

**Comment by josep-m-carbonell** [comment]
The SolversApp is not a replacement of the ExternalSolversApp. It is not the objective, however it has the possibility for including SuperLU and Feast compilation for a particular solver definitions. It was done to test different options for those libraries. Testing SuperLu5.2 and SuperLuMT3.1.  After testing them there is no improvement in speed compared with SuperLU 4.3 (which is included in ExternalSolversApp). Therefore I prefer to not touch ExternalSolverApp, because no improvement would be added.

**Comment by josep-m-carbonell** [comment]
@loumalouomega, this is the same as not looking at the files of the /external_libraries folder of the application. You don't have to review those files.

**Comment by loumalouomega** [comment]
In any case I think that it doesn't make any sense to create a new application. The case of the EigenSolverApplication is special because a very specific library. But for example, if you want to update FEAST you can do a PR updating the version in the ExternalSolversApplication, otherwise you are duplicating unnecessarily the code, and it makes more difficult to maintain in the future. Additionally it will make more difficult to new users to understand how to work with Kratos. Which solver I am supposed to use?, which application I am supposed to compile?, why?Why there are two FEASTs and 3 SuperLU (one of them must be removed because legacy reasons)

**Comment by josep-m-carbonell** [comment]
@loumalouomega I am the maintainer of the new application. Any new user does not have to worry about the dependent applications, there is a work in progress focused on that.

**Comment by loumalouomega** [comment]
> @loumalouomega I am the maintainer of the new application. Any new user does not have to worry about the dependent applications, there is a work in progress focused on that. 

That doesn't justify why creating a new application. When I do a new development I try to do it the most general possible and the most accessible to everyone.

I can understand the creation of a PetSC application, because is a whole library. But duplicating SuperLU and FEAST is not a very clever move in the long run.

File: applications/SolversApplication/linear_system/linear_solvers/feast_solver.hpp [hunk 1]
```diff
@@ -0,0 +1,531 @@
+//
+//   Project Name:        KratosSolversApplication $
```
Comment (loumalouomega) [inline applications/SolversApplication/linear_system/linear_solvers/feast_solver.hpp:2]: @josep-m-carbonell For example this is literally the duplication of existing code, if an error is found in here or in the original one the code should be corrected in two different places, this is against the principals of code development

File: applications/SolversApplication/linear_system/system_builders/system_builder_and_solver.hpp [hunk 2]
```diff
@@ -0,0 +1,545 @@
...
+
+
+/** @brief System Buider and Solver base class
+ *  @details This is the base class for the building and solving the system
+ */
+template<class TSparseSpace,
+         class TDenseSpace, // = DenseSpace<double>,
+         class TLinearSolver //= LinearSolver<TSparseSpace,TDenseSpace>
+         >
+class SystemBuilderAndSolver : public Flags
```
Comment (loumalouomega) [inline applications/SolversApplication/linear_system/system_builders/system_builder_and_solver.hpp:62]: Why creating a new class thst is exactly the same that the BuilderAndSolver class?. If anything relevant is changed in the elements or conditions is changed all this new classes will need to de recoded again

**Comment by RiccardoRossi** [comment]
@josep-m-carbonell if the purpose is testing something than the mechanism would be to have a branch and to eventually discard it if it is found not to be useful.

In particular, the goal of the ExternalSolvers application is to centralize the solver settings so that they can be used by everyone using the Kratos. For this reason an improvement of the FeastSolver or of the SuperLU, which could potentially benefit everyone should go in there so to maximize benefit.
I really think that as @philbucher commented we really cannot afford to duplicate this kind of stuff, since this would further slow done our (already slow) CI.

**Comment by pooyan-dadvand** [comment]
> If you don't want or you don't have time to review it, some other people will do it. The PR can not be splitted.

It can be splitted easily:
* Each added library as a separate PR
* Each added interface for the library as separate PR
* Having the rest of the changes separated by category

**Comment by maceligueta** [comment]
@josep-m-carbonell , in this new application called SolversApplication I see that you are mixing linear solvers with mechanical solvers. Are you sure you don't want to keep what is solid mechanics in the Solid Mechanics Application and what is math in the External Solvers application? Honestly, this refactoring (again) looks very strange to me.

**Comment by josep-m-carbonell** [comment]
It is a good point of view @maceligueta. I will think about it for a future refactoring.

If the problem is the appveyor. We can take away some applications. I will talk with @roigcarlo to disable them.

**Comment by josep-m-carbonell** [comment]
Closing PR to do a smaller one

**Pull request closed without merging**